components/service-operator/apis/storage/v1beta1/s3_types_test.go

```diff
@@ -0,0 +1,157 @@
+
+var _ = Describe("S3Bucket", func() {
+
+	var o v1beta1.S3Bucket
+	var tags []cloudformation.Tag
+
+	BeforeEach(func() {
+		os.Setenv("CLUSTER_NAME", "xxx") // required for env package
+		o = v1beta1.S3Bucket{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "example",
+				Namespace: "default",
+				Labels: map[string]string{
+					cloudformation.AccessGroupLabel: "test.access.group",
+				},
+			},
+			Spec: v1beta1.S3BucketSpec{},
+		}
+		tags = []cloudformation.Tag{
+			{Key: "Cluster", Value: env.ClusterName()},
+			{Key: "Service", Value: "s3"},
+			{Key: "Name", Value: "example"},
+			{Key: "Namespace", Value: "default"},
+			{Key: "Environment", Value: "default"},
+		}
+	})
+
+	It("should default secret name to object name", func() {
+		Expect(o.GetSecretName()).To(Equal("example"))
+	})
+
+	It("should use secret name from spec.Secret if set ", func() {
+		o.Spec.Secret = "my-target-secret"
+		Expect(o.GetSecretName()).To(Equal("my-target-secret"))
+	})
+
+	It("implements runtime.Object", func() {
+		o2 := o.DeepCopyObject()
+		Expect(o2).ToNot(BeZero())
+	})
+
+	Context("cloudformation", func() {
+
+		It("should generate a unique stack name prefixed with cluster name", func() {
+			Expect(o.GetStackName()).To(HavePrefix("xxx-s3-default-example"))
+		})
+
+		It("should require an IAM role input", func() {
+			t := o.GetStackTemplate()
+			Expect(t.Parameters).To(HaveKey("IAMRoleName"))
+		})
+
+		It("should have outputs for connection details", func() {
+			t := o.GetStackTemplate()
+			Expect(t.Outputs).To(HaveKey("S3BucketName"))
+			Expect(t.Outputs).To(HaveKey("IAMRoleName"))
+		})
+
+		It("should map role name to role parameter", func() {
+			params, err := o.GetStackRoleParameters("fake-role")
+			Expect(err).ToNot(HaveOccurred())
+			Expect(params).To(ContainElement(&cloudformation.Parameter{
+				ParameterKey:   aws.String("IAMRoleName"),
+				ParameterValue: aws.String("fake-role"),
+			}))
+		})
+
+		Context("bucket resource", func() {
+
+			var bucket *cloudformation.AWSS3Bucket
+
+			JustBeforeEach(func() {
+				t := o.GetStackTemplate()
+				Expect(t.Resources).To(ContainElement(BeAssignableToTypeOf(&cloudformation.AWSS3Bucket{})))
+				var ok bool
+				bucket, ok = t.Resources[v1beta1.S3BucketResourceName].(*cloudformation.AWSS3Bucket)
+				Expect(ok).To(BeTrue())
+			})
+
+			It("should have a bucket name prefixed with cluster and namespace name", func() {
+				Expect(bucket.BucketName).To(Equal("xxx-default-example"))
+			})
+
+			It("should have suitable tags set", func() {
+				Expect(bucket.Tags).To(Equal(tags))
+			})
+		})
+
+		Context("policy resource", func() {
+			var policy *cloudformation.AWSIAMPolicy
+			var doc cloudformation.PolicyDocument
+
+			JustBeforeEach(func() {
+				t := o.GetStackTemplate()
+				Expect(t.Resources[v1beta1.S3BucketResourceIAMPolicy]).To(BeAssignableToTypeOf(&cloudformation.AWSIAMPolicy{}))
+				policy = t.Resources[v1beta1.S3BucketResourceIAMPolicy].(*cloudformation.AWSIAMPolicy)
```

Comment:
It has `.(*cloudformation.AWSIAMPolicy)`. What does this dot operator do?

Reply:
If I'm honest, I don't know. But all the tests pass:

```
S3Bucket 
  should default secret name to object name
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:43
•
------------------------------
S3Bucket 
  should use secret name from spec.Secret if set 
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:47
•
------------------------------
S3Bucket 
  implements runtime.Object
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:52
•
------------------------------
S3Bucket cloudformation 
  should generate a unique stack name prefixed with cluster name
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:59
•
------------------------------
S3Bucket cloudformation 
  should require an IAM role input
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:63
•
------------------------------
S3Bucket cloudformation 
  should have outputs for connection details
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:68
•
------------------------------
S3Bucket cloudformation 
  should map role name to role parameter
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:74
•
------------------------------
S3Bucket cloudformation bucket resource 
  should have a bucket name prefixed with cluster and namespace name
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:95
•
------------------------------
S3Bucket cloudformation bucket resource 
  should have suitable tags set
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:99
•
------------------------------
S3Bucket cloudformation policy resource 
  should have a policy name
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:116
•
------------------------------
S3Bucket cloudformation policy resource 
  should asign policy to the given role name
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:120
•
------------------------------
S3Bucket cloudformation policy resource 
  should have a policy document with relevant actions
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:124
•
------------------------------
S3Bucket cloudformation policy resource 
  is scoped to the correct resources (a single S3 bucket and its objects)
  /home/isabelllong/repos/gds/gsp/gsp/components/service-operator/apis/storage/v1beta1/s3_types_test.go:132
•

Ran 13 of 13 Specs in 0.003 seconds
SUCCESS! -- 13 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
coverage: 48.3% of statements
```

Reply:
that is a cast from the `interface{}` (anything) type to the type `*cloudformation.AWSIAMPolicy` ... a cast in this way will panic if it is unable to perform the cast (but we check it can be cast in the line before)